### ISSUE TITLE
Lint request: unbuffered I/O usage

### ISSUE THREAD

**Issue by olson-dan** [Manishearth/rust-clippy#1805]
I and several other beginners have run into the surprising issue that Rust's default file reading and writing methods are unbuffered and buffered IO is purely opt-in with this language.  It may be useful to add a lint to warn users about this, as it seems to come up from time to time when comparing programs to other languages such as Python.

**Comment by clarcharr** [comment]
This shouldn't apply when methods like `read_to_end` or `read_to_string` are used, because no intermediate buffering is necessary in that case.

**Comment by olson-dan** [comment]
Yes, I agree, the lint would need to indicate unbuffered IO as an issue only when there is a likely performance issue.

**Comment by llogiq** [comment]
The lint would trigger on all eligible Read/Write operations where `self` is a `File`/`Stdin`/`Stdout`/`StdinLock`/`StdoutLock`/`TcpStream`/`UnixStream`. Suggest wrapping `self` in a `BufReader`/`BufWriter` to improve performance.

Note that there will be false positives where a) the direct IO is really intentional or b) a custom buffering solution is taken. Those are unlikely to be hit by new users, and one can deactivate this lint on a case-by-case basis. The docs should reflect this.

**Comment by pickfire** [comment]
Note that `StdinLock` is buffered. I will open another issue for this and it could be specifically for `BufReader::new(io::stdin())`.
